runtime/major_gc.c

```diff
@@ -749,12 +757,32 @@ static uintnat gc_slice_work(uintnat heap_words,
 
       return max2 (alloc_work, offheap_work);
     }
-  case GC_PACING_2025_FULL: /* TODO: separate this out. */
-    /* fall through */
-  case GC_PACING_2025_SIMPLE: {
+  case GC_PACING_2025: {
     /* Shiny new 2025 Doligez/Dolan pacing policy */
     double sweep_per_mark = (double)caml_percent_sweep_per_mark / 100.0;
     double space_overhead = (double)caml_percent_free / 100.0;
+    if (caml_gc_overhead_adjustment != 0) {
+      /* In theory, we should adjust according to the allocation rate.
+
+         However, this produces a dependence on real time, making the GC
+         nondeterministic even for single-threaded programs and doing weird
+         things for programs which pause in I/O for a long time.
+
+         Instead, we use minor heap allocation as a proxy for time, so instead
+         of measuring the allocation rate we measure the promotion rate.
+
+         The promotion rate is scaled so that the nominal promotion rate of 10%
+         comes out as 1.0, and direct-to-major allocations are deemed to have
+         this nominal promotion rate. */
+      double scaled_prom_rate =
+        (10.0 * allocated_words) /
+        (10.0 * allocated_direct_words + minor_words);
```

Comment:
Can this divide by zero?

Reply:
Hmm, probably. (Hard to trigger this code without doing some allocation, so maybe not. Worth fixing regardless)